pkg/component/watch.go

```diff
@@ -0,0 +1,166 @@
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			folders = append(folders, newPath)
+		}
+		return nil
+	})
+	for _, v := range folders {
+		ignore := false
+		for _, pattern := range ignores {
+			if matched, _ := regexp.MatchString(pattern, v); matched {
+				ignore = true
+				break
+			}
+		}
+		if ignore {
+			log.Debugf("ignoring watch for %s", v)
+			continue
+		}
+		log.Debugf("adding watch on path %s", v)
+		err = watcher.Add(v)
+		if err != nil {
+			// Linux "no space left on device" issues are usually resolved via
+			// $ sudo sysctl fs.inotify.max_user_watches=65536
+			// BSD / OSX: "too many open files" issues are ussualy resolved via
+			// $ sysctl variables "kern.maxfiles" and "kern.maxfilesperproc",
+			return fmt.Errorf("error adding watcher for path %s: %v", v, err)
+		}
+	}
+	return nil
+}
+
+// WatchAndPush watched directory dir, if something changes in that directory calls push
+// ignores .git/* by default
+// inspired by https://github.com/openshift/origin/blob/e785f76194c57bd0e1674c2f2776333e1e0e4e78/pkg/oc/cli/cmd/rsync/rsync.go#L257
+func WatchAndPush(client *occlient.Client, componentName string, dir string, out io.Writer) error {
+	log.Debugf("starting WatchAndPush, dir: %s, component: %s", dir, componentName)
+
+	// it might be better to expose this as argument in the future
+	ignores := []string{
+		// ignore git as it can change even if no source file changed
+		// for example some plugins providing git info in PS1 doing that
+		".*\\.git.*",
+	}
+
+	// these variables must be accessed while holding the changeLock
+	// mutex as they are shared between goroutines to communicate
+	// sync state/events.
+	var (
+		changeLock   sync.Mutex
+		dirty        bool
+		lastChange   time.Time
+		watchError   error
+		changedFiles []string
+	)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error setting up filesystem watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				changeLock.Lock()
+				log.Debugf("filesystem watch event: %s", event)
+				changedFiles = append(changedFiles, event.Name)
+				lastChange = time.Now()
+				dirty = true
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					if e := watcher.Remove(event.Name); e != nil {
+						log.Debugf("error removing watch for %s: %v", event.Name, e)
+					}
+				} else {
+					if e := addRecursiveWatch(watcher, event.Name, ignores); e != nil && watchError == nil {
+						watchError = e
+					}
+				}
+				changeLock.Unlock()
+			case err := <-watcher.Errors:
+				changeLock.Lock()
+				watchError = fmt.Errorf("error watching filesystem for changes: %v", err)
+				changeLock.Unlock()
+			}
+		}
+	}()
+	err = addRecursiveWatch(watcher, dir, ignores)
+	if err != nil {
+		return fmt.Errorf("error watching source path %s: %v", dir, err)
+	}
+
+	delay := 5 * time.Second
```

Comment:
should be a global variable

Reply:
why global if it is not used anywhere else?